### PR TITLE
[AutoDiff] Handle `@differentiable(linear)` on Types

### PR DESCRIPTION
**Pull request by bartchr808**
PR adds support to mark types as `@differentiable(linear)`, serialize/deserialize it, and print it in the AST when you run `-dump-ast`. However, currently if you mark a type as differentiable, I throw a diagnostic error. I added very few tests, as there aren't a lot of cases to consider.

### Questions/Comments
- Because of us currently creating the error diagnostic, I don't think I can test serialization since an error would be thrown.

### CODE REVIEW

File: include/swift/Serialization/ModuleFormat.h [hunk 1]
```diff
@@ -1662,6 +1664,7 @@ namespace decls_block {
   >;
 
   // SWIFT_ENABLE_TENSORFLOW
+  // TODO: why didn't I add a bit for linear flag here?
```
Comment (bartchr808) [inline include/swift/Serialization/ModuleFormat.h]: Forgot to do this in previous PR, will add it in this PR and add corresponding tests.

File: include/swift/AST/Types.h [hunk 2]
```diff
@@ -1756,6 +1756,7 @@ class ParameterTypeFlags {
     OwnershipShift = 2,
     Ownership   = 7 << OwnershipShift,
     // SWIFT_ENABLE_TENSORFLOW
+    // TODO: need to handle parameters
```
Comment (bartchr808) [inline include/swift/AST/Types.h]: Will do this in a separate PR

File: include/swift/AST/Types.h [hunk 3]
```diff
@@ -2899,17 +2901,28 @@ class AnyFunctionType : public TypeBase {
...
-    bool isDifferentiable() const { return Bits & DifferentiableMask; }
+    bool isDifferentiable() const {
+      return ((DifferentiabilityKind)DifferentiableMask ==
+              DifferentiabilityKind::Normal) || isLinear();
+    }
+    bool isLinear() const {
+      return ((DifferentiabilityKind)DifferentiableMask ==
+              DifferentiabilityKind::Linear);
+    }
+    DifferentiabilityKind getDifferentiableMask() const {
```
Comment (rxwei) [inline include/swift/AST/Types.h]: Hmm, this is incorrect. A mask has no meaning. It's just some set bits in a random spot. It should never be coerced into a `DifferentiabilityKind`. Instead, we should define a `getDifferentiabilityKind()` and return `DifferentiabilityMask(DifferentiableMask << DifferentiableOffset)`.

File: include/swift/AST/Types.h [hunk 4]
```diff
@@ -2899,17 +2901,28 @@ class AnyFunctionType : public TypeBase {
...
 
     bool isNoEscape() const { return Bits & NoEscapeMask; }
     bool throws() const { return Bits & ThrowsMask; }
     // SWIFT_ENABLE_TENSORFLOW
-    bool isDifferentiable() const { return Bits & DifferentiableMask; }
+    bool isDifferentiable() const {
+      return ((DifferentiabilityKind)DifferentiableMask ==
+              DifferentiabilityKind::Normal) || isLinear();
+    }
+    bool isLinear() const {
```
Comment (rxwei) [inline include/swift/AST/Types.h]: This method isn't necessary. All we need is a `getDifferentiabilityKind()` and `isDifferentiable()` (implemented in terms of `getDifferentiabilityKind()`).

File: include/swift/AST/Types.h [hunk 5]
```diff
@@ -2899,17 +2901,28 @@ class AnyFunctionType : public TypeBase {
...
+      Bits |=
+          ((unsigned)diffKind << DifferentiableMaskOffset) & DifferentiableMask;
     }
 
     bool isNoEscape() const { return Bits & NoEscapeMask; }
     bool throws() const { return Bits & ThrowsMask; }
     // SWIFT_ENABLE_TENSORFLOW
-    bool isDifferentiable() const { return Bits & DifferentiableMask; }
+    bool isDifferentiable() const {
+      return ((DifferentiabilityKind)DifferentiableMask ==
```
Comment (rxwei) [inline include/swift/AST/Types.h]:
```suggestion
      return getDifferentiabilityKind() >= Normal;
```

File: include/swift/AST/Types.h (hunk 3 above)
Comment (bartchr808) [reply include/swift/AST/Types.h]: Ya this seems like the one bug I had, but now I have another one that I'm looking into. And just making sure shouldn't it be `DifferentiableMask >> DifferentiableOffset` since we want to undo the offset?

Comment (rxwei) [reply include/swift/AST/Types.h]: Sorry, please disregard my earlier suggestion.

Also, I'd suggest doing the following:
* Rename `DifferentiableMask` to `DifferentiabilityMask` and `DifferentiableMaskOffset` to `DifferentiabilityMaskOffset`, since this represents something more than just a binary flag.
* Add a `NumDifferentiabilityMaskBits = 3` right below `DifferentiableMask` in the enum.
* In `getDifferentiabilityKind()`, return `DifferentiabilityKind((Bits & DifferentiableMask) >> DifferentiableMaskOffset)`.

Comment (bartchr808) [reply include/swift/AST/Types.h]: Shouldn't `NumDifferentiabilityMaskBits = 2` since `3=0x3=0b11` so it is 2 bits?

Comment (rxwei) [reply include/swift/AST/Types.h]: Ah yes, I had a typo.

Comment (bartchr808) [reply include/swift/AST/Types.h]: What do we need `NumDifferentiabilityMaskBits` for? For `NumMaskBits` I see it is used for the following assert:
```c++
static_assert(ExtInfo::NumMaskBits == NumAFTExtInfoBits,
              "ExtInfo and AnyFunctionTypeBitfields must agree on bit size");
```
Should something like this be done for `NumDifferentiabilityMaskBits` as well?

Comment (rxwei) [reply include/swift/AST/Types.h]: Something asserting that the mask, offset, and num-bits are consistent would be good, but I'm not sure if something like that can be statically evaluable.

File: include/swift/AST/Types.h [hunk 6]
```diff
@@ -2985,16 +2986,11 @@ class AnyFunctionType : public TypeBase {
...
     const {
-      switch (differentiability) {
-        case DifferentiabilityKind::NonDifferentiable:
-          return ExtInfo(Bits & ~DifferentiabilityMask);
-        case DifferentiabilityKind::Normal:
-          return ExtInfo(Bits & ~(0b10 << DifferentiabilityMaskOffset));
-        case DifferentiabilityKind::Linear:
-          return ExtInfo(Bits | DifferentiabilityMask);
-      }
+      unsigned clearedBits = Bits & ~(0b11 << DifferentiabilityMaskOffset);
```
Comment (rxwei) [inline include/swift/AST/Types.h]: `(0b11 << DifferentiabilityMaskOffset)` is just equivalent to `DifferentiabilityMask`, so we should use that instead. With that, this function can be implemented with one return statement.

File: lib/AST/ASTPrinter.cpp [hunk 7]
```diff
@@ -3806,7 +3806,11 @@ class TypePrinter : public TypeVisitor<TypePrinter> {
     // SWIFT_ENABLE_TENSORFLOW
     if (!Options.excludeAttrKind(TAK_differentiable) && info.isDifferentiable()) {
       // FIXME(rxwei): Print differentiation order.
-      Printer << "@differentiable ";
+      if (info.getDifferentiabilityKind() == DifferentiabilityKind::Linear) {
```
Comment (bartchr808) [inline lib/AST/ASTPrinter.cpp]: Forgot to do this for func decls, will do this in another PR

File: include/swift/AST/DiagnosticsParse.def [hunk 8]
```diff
@@ -1500,9 +1500,8 @@ ERROR(attr_differentiable_missing_label,PointsToFirstBadToken,
 ERROR(attr_differentiable_expected_label,none,
       "expected either 'wrt:' or a function specifier label, e.g. 'jvp:', "
       "or 'vjp:'", ())
-ERROR(attr_differentiable_no_vjp_or_jvp_when_linear,none,
```
Comment (bartchr808) [inline include/swift/AST/DiagnosticsParse.def]: This was in the wrong file, it's not a parse error but a sema error

File: include/swift/Serialization/ModuleFormat.h [hunk 9]
```diff
@@ -851,7 +851,7 @@ namespace decls_block {
     FunctionTypeRepresentationField, // representation
     BCFixed<1>,          // throws?
     // SWIFT_ENABLE_TENSORFLOW
-    BCFixed<1>,          // differentiable?
+    BCFixed<2>,          // differentiable & linear?
```
Comment (bartchr808) [inline include/swift/Serialization/ModuleFormat.h]: For Generic Function Type, this and also [this line](https://github.com/apple/swift/pull/25325/files#diff-5fa18b91a026a0b29b5c3b4d59feaf95R4728) were the only thing I updated. I'm not sure whether I need to update anymore.

File: test/AutoDiff/differentiable_func_type_parse.swift [hunk 10]
```diff
@@ -0,0 +1,5 @@
+// RUN: %target-swift-frontend -parse -verify %s
+
+let a: @differentiable (Float) -> Float // okay
+
+let b: @differentiable(linear) (Float) -> Float // okay
```
Comment (bartchr808) [inline test/AutoDiff/differentiable_func_type_parse.swift]: From what I saw, no parsing tests were done for marking types as differentiable so I created a new file.

Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_parse.swift]: I am trying to add this test:
```swift
// expected-error @+2 {{expected ')' after 'linear' in '@differentiable' attribute}}
// expected-note @+1 {{to match this opening '('}}
let c: @differentiable(linear (Float) -> Float
```
But am getting an error:
```
differentiable_func_type_parse.swift:8:4: error: expected note not produced
// expected-note @+1 {{to match this opening '('}}
~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<unknown>:0: error: unexpected note produced: to match this opening '('
```

Comment (rxwei) [reply test/AutoDiff/differentiable_func_type_parse.swift]: Maybe the source location is wrong? Try running it in a terminal and see the source location.

File: include/swift/Serialization/ModuleFormat.h (hunk 9 above)
Comment (rxwei) [reply include/swift/Serialization/ModuleFormat.h]: The comment should be changed to `differentiability kind`.

File: test/AutoDiff/differentiable_func_type_parse.swift (hunk 10 above)
Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_parse.swift]: Ahh yes so it actually prints out differently then I was expecting. I included the second error below, however that is a type check error, not a parse error which this test doesn't see.

```swift
diff.swift:18:31: error: expected ')' after 'linear' in '@differentiable' attribute
let c: @differentiable(linear (Float) -> Float
                              ^
<unknown>:0: note: to match this opening '('
diff.swift:18:9: error: cannot mark types as linear differentiable
let c: @differentiable(linear (Float) -> Float
        ^
```

Comment (rxwei) [reply test/AutoDiff/differentiable_func_type_parse.swift]: It should not emit a type error when there's a parse error. It means that you probably didn't `return true` or something after you emitted a parser diagnostic.

File: include/swift/AST/Attr.h [hunk 11]
```diff
@@ -67,6 +67,11 @@ class TypeAttributes {
   SourceLoc AtLoc;
   Optional<StringRef> convention = None;
   Optional<StringRef> conventionWitnessMethodProtocol = None;
+  
+  // SWIFT_ENABLE_TENSORFLOW
+  // indicates whether the type's '@differentiable' attribute has a 'linear'
```
Comment (rxwei) [inline include/swift/AST/Attr.h]:
```suggestion
  // Indicates whether the type's '@differentiable' attribute has a 'linear'
```

File: include/swift/AST/Attr.h [hunk 12]
```diff
@@ -67,6 +67,11 @@ class TypeAttributes {
...
+  // argument
```
Comment (rxwei) [inline include/swift/AST/Attr.h]:
```suggestion
  // argument.
```

File: include/swift/AST/Types.h [hunk 13]
```diff
@@ -3854,11 +3876,12 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
...
     }
     // SWIFT_ENABLE_TENSORFLOW
-    ExtInfo withDifferentiable(bool isDifferentiable = true) const {
-      if (isDifferentiable)
-        return ExtInfo(Bits | DifferentiableMask);
-      else
-        return ExtInfo(Bits & ~DifferentiableMask);
+    ExtInfo withDifferentiabilityKind(
+        DifferentiabilityKind differentiability)
+    const {
```
Comment (rxwei) [inline include/swift/AST/Types.h]: Move this to the end of the previous line

File: include/swift/Serialization/ModuleFormat.h [hunk 14]
```diff
@@ -777,16 +777,16 @@ namespace decls_block {
     BCFixed<1>,  // noescape?
     // SWIFT_ENABLE_TENSORFLOW
     BCFixed<1>,  // throws?
-    BCFixed<1>   // differentiable?
+    BCFixed<2> // differentiable & linear?
```
Comment (rxwei) [inline include/swift/Serialization/ModuleFormat.h]:
```suggestion
    BCFixed<2> // differentiability kind
```

File: lib/Parse/ParseDecl.cpp [hunk 15]
```diff
@@ -2395,6 +2395,28 @@ bool Parser::parseTypeAttribute(TypeAttributes &Attributes, bool justChecking) {
                        diag::convention_attribute_expected_rparen,
                        LPLoc);
   }
+  
+  // SWIFT_ENABLE_TENSORFLOW
+  bool linear = false;
+  if (attr == TAK_differentiable) {
+    // Check if there is a 'linear' argument.
+    if (Tok.is(tok::l_paren) && peekToken().is(tok::identifier) && peekToken().getText() == "linear") {
```
Comment (rxwei) [inline lib/Parse/ParseDecl.cpp]: Seems to be over 80 columns.

File: lib/SIL/SILFunctionType.cpp [hunk 16]
```diff
@@ -124,7 +124,8 @@ CanSILFunctionType SILFunctionType::getWithDifferentiability(
                 : SILParameterDifferentiability::NotDifferentiable));
   }
 
-  auto newExtInfo = getExtInfo().withDifferentiable();
+  auto newExtInfo = getExtInfo().withDifferentiabilityKind(DifferentiabilityKind
```
Comment (rxwei) [inline lib/SIL/SILFunctionType.cpp]:
```suggestion
  auto newExtInfo = getExtInfo().withDifferentiabilityKind(
      DifferentiabilityKind::Normal);
```

File: lib/SIL/SILFunctionType.cpp [hunk 17]
```diff
@@ -135,7 +136,8 @@ CanSILFunctionType SILFunctionType::getWithDifferentiability(
 CanSILFunctionType SILFunctionType::getWithoutDifferentiability() {
   if (!isDifferentiable())
     return CanSILFunctionType(this);
-  auto nondiffExtInfo = getExtInfo().withDifferentiable(false);
+  auto nondiffExtInfo = getExtInfo().withDifferentiabilityKind(
+                            DifferentiabilityKind::NonDifferentiable);
```
Comment (rxwei) [inline lib/SIL/SILFunctionType.cpp]:
```suggestion
        DifferentiabilityKind::NonDifferentiable);
```

File: lib/Sema/CSApply.cpp [hunk 18]
```diff
@@ -6495,7 +6495,7 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
 
     // SWIFT_ENABLE_TENSORFLOW
     auto toEINoADConversion =
-        toEI.withDifferentiable(fromEI.isDifferentiable());
+    toEI.withDifferentiabilityKind(fromEI.getDifferentiabilityKind());
```
Comment (rxwei) [inline lib/Sema/CSApply.cpp]:
```suggestion
        toEI.withDifferentiabilityKind(fromEI.getDifferentiabilityKind());
```

File: lib/Sema/TypeCheckType.cpp [hunk 19]
```diff
@@ -2225,12 +2225,21 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
               SourceLoc(), Context.getIdentifier(protocolName));
         }
       }
+      
+      DifferentiabilityKind diffkind = DifferentiabilityKind::NonDifferentiable;
+      if (attrs.has(TAK_differentiable)) {
+        if (attrs.linear) {
+          diffkind = DifferentiabilityKind::Linear;
```
Comment (rxwei) [inline lib/Sema/TypeCheckType.cpp]: Use a ternary here?

File: lib/Sema/TypeCheckType.cpp [hunk 20]
```diff
@@ -2264,12 +2273,25 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
                  "@autoclosure");
         attrs.clearAttribute(TAK_autoclosure);
       }
-
+
```
Comment (rxwei) [inline lib/Sema/TypeCheckType.cpp]: // SWIFT_ENABLE_TENSORFLOW

File: lib/Sema/TypeCheckType.cpp [hunk 21]
```diff
@@ -2264,12 +2273,25 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
...
       }
-
+      
+      DifferentiabilityKind diffkind = DifferentiabilityKind::NonDifferentiable;
+      if (attrs.has(TAK_differentiable)) {
+        if (attrs.linear) {
+          // FIXME(bartchr): allow types to be marked
+          // as `@differentiable(linear)`.
+          diagnose(attrs.getLoc(TAK_differentiable),
+                   diag::linear_differentiable_type_disabled);
```
Comment (rxwei) [inline lib/Sema/TypeCheckType.cpp]:
```suggestion
                   diag::linear_differentiable_type_disabled);
          attrs.clearAttribute(TAK_differentiable);
```

File: lib/Sema/TypeCheckType.cpp [hunk 22]
```diff
@@ -2264,12 +2273,25 @@ Type TypeResolver::resolveAttributedType(TypeAttributes &attrs,
...
+          diffkind = DifferentiabilityKind::Linear;
```
Comment (rxwei) [inline lib/Sema/TypeCheckType.cpp]:
```suggestion
```

File: lib/Sema/TypeCheckType.cpp (hunk 21 above)
Comment (rxwei) [reply lib/Sema/TypeCheckType.cpp]: Since this is an error, we should clear the attribute as if it never existed and move on, like how Swift deals with the invalid `TAK_autoclosure` above.

File: lib/Serialization/Deserialization.cpp [hunk 23]
```diff
@@ -4704,7 +4704,9 @@ class swift::TypeDeserializer {
     uint8_t rawRepresentation;
 
     // SWIFT_ENABLE_TENSORFLOW
-    bool noescape = false, throws = false, differentiable = false;
+    bool noescape = false, throws = false;
+    uint8_t rawDiffKind;
+    DifferentiabilityKind diffKind = DifferentiabilityKind::NonDifferentiable;
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]:
```suggestion
    auto diffKind = DifferentiabilityKind::NonDifferentiable;
```

File: lib/Serialization/Deserialization.cpp [hunk 24]
```diff
@@ -4704,7 +4704,9 @@ class swift::TypeDeserializer {
     uint8_t rawRepresentation;
 
     // SWIFT_ENABLE_TENSORFLOW
-    bool noescape = false, throws = false, differentiable = false;
+    bool noescape = false, throws = false;
+    uint8_t rawDiffKind;
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]:
```suggestion
    uint8_t rawDiffKind = 0;
```

Comment (rxwei) [reply lib/Serialization/Deserialization.cpp]: Initialize values to prevent undefined behavior during debugging.

File: lib/Serialization/Deserialization.cpp [hunk 25]
```diff
@@ -4713,17 +4715,20 @@ class swift::TypeDeserializer {
                                                   noescape,
                                                   // SWIFT_ENABLE_TENSORFLOW
                                                   throws,
-                                                  differentiable);
+                                                  rawDiffKind);
+      diffKind = DifferentiabilityKind(rawDiffKind);
+
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]:
```suggestion
```

File: lib/Serialization/Deserialization.cpp [hunk 26]
```diff
@@ -4713,17 +4715,20 @@ class swift::TypeDeserializer {
...
       decls_block::GenericFunctionTypeLayout::readRecord(scratch,
                                                          resultID,
                                                          rawRepresentation,
                                                          throws,
                                                          // SWIFT_ENABLE_TENSORFLOW
-                                                         differentiable,
+                                                         rawDiffKind,
                                                          rawGenericSig);
       genericSig = MF.getGenericSignature(rawGenericSig);
+      diffKind = DifferentiabilityKind(rawDiffKind);
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]:
```suggestion
      // SWIFT_ENABLE_TENSORFLOW
      diffKind = DifferentiabilityKind(rawDiffKind);
```

File: lib/Serialization/Deserialization.cpp [hunk 27]
```diff
@@ -5112,8 +5117,10 @@ class swift::TypeDeserializer {
       return nullptr;
     }
     // SWIFT_ENABLE_TENSORFLOW
-    SILFunctionType::ExtInfo extInfo(*representation, pseudogeneric, noescape,
-                                     differentiable);
+    DifferentiabilityKind kind =
```
Comment (rxwei) [inline lib/Serialization/Deserialization.cpp]:
```suggestion
    auto kind =
```

File: test/AutoDiff/differentiable_func_type_type_checking.swift [hunk 28]
```diff
@@ -63,3 +63,10 @@ let diffVectorFunc: (Vector<Float>) -> Vector<Float>
 inferredConformancesGeneric(diffVectorFunc) // okay!
 
 func inferredConformancesGenericResult<T, U>() -> @differentiable (Vector<T>) -> Vector<U> {}
+
+//
+// linear function type
+//
+
+// expected-error @+1 {{cannot mark types as linear differentiable}}
+let _: @differentiable(linear) (Float) -> Float
```
Comment (rxwei) [inline test/AutoDiff/differentiable_func_type_type_checking.swift]: I'd suggest adding more ambiguity tests because parsing is tricky. See https://github.com/rxwei/swift/blob/ffc1370297eb2903b3a6ab2254c84c8ecf616bf1/test/AutoDiff/differentiable_func_type_parse.swift.

Comment (rxwei) [reply test/AutoDiff/differentiable_func_type_type_checking.swift]: I meant to attach this comment to the parsing test file.

File: test/AutoDiff/differentiable_func_type_parse.swift (hunk 10 above)
Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_parse.swift]: Yes now with the new way I handle the token with the backtrack so I can look ahead multiple tokens, I think this problem went away. Adding more test cases because of the change as well.

File: test/AutoDiff/differentiable_func_type_type_checking.swift (hunk 28 above)
Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_type_checking.swift]: Similar to my comment above regarding adding the back tracking, I'm now better able to handle these test cases so I added several of them. 😄

File: lib/Sema/TypeCheckType.cpp (hunk 19 above)
Comment (bartchr808) [reply lib/Sema/TypeCheckType.cpp]: Ahh yes I simplified my nested ternary too much! Before I had the double nested ternary, but now I can just have one.

File: lib/Serialization/Deserialization.cpp (hunk 25 above)
Comment (bartchr808) [reply lib/Serialization/Deserialization.cpp]: Ahh how were you able to tell there was extra whitespace there?

File: lib/Serialization/Deserialization.cpp (hunk 26 above)
Comment (bartchr808) [reply lib/Serialization/Deserialization.cpp]: Should I put one on [line 4719](https://github.com/apple/swift/pull/25325/files/c3b6c8133bf2e8f8d06d9c3e7ae0077469c42ec5#diff-5fa18b91a026a0b29b5c3b4d59feaf95R4719) in `Deserialzation.cpp` (this same file) as well?

File: lib/Serialization/Deserialization.cpp (hunk 25 above)
Comment (rxwei) [reply lib/Serialization/Deserialization.cpp]: Hmm, it's actually not about whitespace, but about the extra line.

File: lib/Serialization/Deserialization.cpp (hunk 26 above)
Comment (rxwei) [reply lib/Serialization/Deserialization.cpp]: That place already has one.

File: lib/Serialization/Deserialization.cpp (hunk 25 above)
Comment (bartchr808) [reply lib/Serialization/Deserialization.cpp]: Oh okay will remove it

File: lib/Serialization/Deserialization.cpp (hunk 26 above)
Comment (bartchr808) [reply lib/Serialization/Deserialization.cpp]: Resolved offline: I flipped the order of `genericSig` and `diffKind` so I could remove the flag.

File: lib/Parse/ParseDecl.cpp [hunk 29]
```diff
@@ -2395,6 +2395,53 @@ bool Parser::parseTypeAttribute(TypeAttributes &Attributes, bool justChecking) {
...
+  if (attr == TAK_differentiable) {
+    // Check if there is a 'linear' argument.
+    if (Tok.is(tok::l_paren) && peekToken().is(tok::identifier)) {
+      Parser::BacktrackingScope backtrack(*this);
+      consumeToken(tok::l_paren);
+      
+      // Determine if we have '@differentiable(linear) (T) -> U'
+      // or '@differentiable (linear) -> U'.
+      if (Tok.getText() == "linear" && consumeIf(tok::identifier)) {
+
```
Comment (rxwei) [inline lib/Parse/ParseDecl.cpp]:
```suggestion
```

File: test/AutoDiff/differentiable_func_type_parse.swift [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+  struct linear : Differentiable {}
+  let property: @differentiable (linear) -> Float // okay
+  let property: @differentiable(linear) (linear) -> linear // okay
+  let property: @differentiable (linear, linear) -> linear // okay
+  let property: @differentiable (linear, Float) -> linear // okay
+  let property: @differentiable (Float, linear) -> linear // okay
+  let property: @differentiable(linear) (linear, linear, Float, linear)
+    -> Float // okay
+  // expected-error @+1 {{expected ')' after 'linear' in '@differentiable' attribute}}
+  let property: @differentiable(linear (linear) -> Float
```
Comment (rxwei) [inline test/AutoDiff/differentiable_func_type_parse.swift]: I'll throw an unlikely case here... Feel free to ignore it.
```swift
typealias linear = (Float) -> Float
let _: @differentiable(linear) // hell yeah
```

**Review by rxwei — APPROVED** [review]
LGTM

File: lib/Parse/ParseDecl.cpp (hunk 29 above)
Comment (bartchr808) [reply lib/Parse/ParseDecl.cpp]: Will remove the extra line this time instead of just the whitespace 😉

File: test/AutoDiff/differentiable_func_type_parse.swift (hunk 30 above)
Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_parse.swift]: Ahh yes with the typealias I'll add that!

Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_parse.swift]: Good catch! As this test made me realize that this works like expected:
```swift
let property: @differentiable linear // okay
```
However, this should work but doesn't:
```swift 
let property: @differentiable(linear) linear // should be okay, but fails
```

```
error: unexpected error produced: consecutive declarations on a line must be separated by ';'
  let property: @differentiable(linear) linear // okay
                                       ^
```

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

File: test/AutoDiff/differentiable_func_type_parse.swift (hunk 30 above)
Comment (bartchr808) [reply test/AutoDiff/differentiable_func_type_parse.swift]: Offline we decided to comment out this test case and file a [JIRA ticket](https://bugs.swift.org/browse/TF-576) about it.

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow

**Comment by bartchr808** [comment]
@swift-ci please test tensorflow
